src/theory/quantifiers/sygus/transition_inference.h

```diff
@@ -0,0 +1,327 @@
+   */
+  bool processDisjunct(Node n,
+                       std::map<bool, Node>& terms,
+                       std::vector<Node>& disjuncts,
+                       std::map<bool, std::map<Node, bool> >& visited,
+                       bool topLevel);
+  /**
+   * This method infers if the conjunction of disjuncts is equivalent to a
+   * conjunction of the form
+   *   (~) const_var[1] = const_subs[1] ... (~) const_var[n] = const_subs[n]
+   * where the above equalities are negated iff reqPol is false, and
+   *   const_var[1] ... const_var[n]
+   * are distinct members of vars
+   */
+  void getConstantSubstitution(const std::vector<Node>& vars,
+                               const std::vector<Node>& disjuncts,
+                               std::vector<Node>& const_var,
+                               std::vector<Node>& const_subs,
+                               bool reqPol);
+  /** get normalized substitution
+   *
+   * This method takes as input a node curr of the form I( t1, ..., tn ) and
+   * a vector of variables pvars, where pvars.size()=n. For each ti that is
+   * a variable, it adds ti to vars, and pvars[i] to subs. For each ti that is
+   * not a variable, it adds the disequality ti != pvars[i] to disjuncts.
+   *
+   * This function is used for instance to normalize an arbitrary application of
+   * I so that is over arguments pvars. For instance if curr is I(3,5,y) and
+   * pvars = { x1,x2,x3 }, then the formula:
+   *   I(3,5,y) ^ P(y)
+   * is equivalent to:
+   *   x1 != 3 V x2 != 5 V I(x1,x2,x3) V P( y ) { y -> x3 }
+   * Here, we add y and x3 to vars and subs respectively, and x1!=3 and x2!=5
+   * to disjuncts.
+   */
+  void getNormalizedSubstitution(Node curr,
+                                 const std::vector<Node>& pvars,
+                                 std::vector<Node>& vars,
+                                 std::vector<Node>& subs,
+                                 std::vector<Node>& disjuncts);
+  /**
+   * Stores one of the components of the inferred form of the synthesis
+   * conjecture (precondition, postcondition, or transition relation).
+   */
+  class Component
+  {
+   public:
+    Component() {}
+    /** The formula that was inferred for this component */
+    Node d_this;
+    /** The list of conjuncts of the above formula */
+    std::vector<Node> d_conjuncts;
+    /**
+     * Maps formulas to he constant equality substitution that it entails.
+     * For example, the formula (x=4 ^ y=x+5) may map to { x -> 4, y -> 9 }.
+     */
+    std::map<Node, std::map<Node, Node> > d_const_eq;
+    /** Does this component have conjunct c? */
+    bool has(Node c) const
+    {
+      return std::find(d_conjuncts.begin(), d_conjuncts.end(), c)
+             != d_conjuncts.end();
+    }
+  };
+  /**
+   * A collection of components, where:
+   * 1 -> precondition, -1 -> postcondition, 0 -> transition relation.
+   */
+  std::map<int, Component> d_com;
```

Comment:
make this a fixed-size array or something

Reply:
Going to switch this to the most obvious thing (3 independent data members) :)
Should have been like this in the first place.

Reply:
Best kind of fixed-size array! :)